src/security/body_parse/body_parsing.cpp

```diff
@@ -0,0 +1,158 @@
+}
+
+bool is_multipart(const ngx_http_request_t &req) {
+  const ngx_table_elt_t *ct = req.headers_in.content_type;
+  return ct && is_content_type(datadog::nginx::to_string_view(ct->value),
+                               "multipart/form-data"sv);
+}
+
+bool is_urlencoded(const ngx_http_request_t &req) {
+  const ngx_table_elt_t *ct = req.headers_in.content_type;
+  return ct && is_content_type(datadog::nginx::to_string_view(ct->value),
+                               "application/x-www-form-urlencoded"sv);
+}
+
+}  // namespace
+
+namespace datadog::nginx::security {
+
+bool parse_body(ddwaf_obj &slot, ngx_http_request_t &req,
+                const ngx_chain_t &chain, std::size_t size,
+                DdwafMemres &memres) {
+  if (is_json(req)) {
+    // use rapidjson to parse:
+    bool success = parse_json(slot, req, chain, memres);
+    if (success) {
+      return true;
+    }
+  }
+
+  if (is_multipart(req)) {
+    std::optional<HttpContentType> ct = HttpContentType::for_string(
+        to_string_view(req.headers_in.content_type->value));
+    if (!ct) {
+      ngx_log_error(NGX_LOG_NOTICE, req.connection->log, 0,
+                    "multipart: invalid multipart/form-data content-type");
+      return false;
+    }
+
+    return parse_multipart(slot, req, *ct, chain, memres);
+  }
+
+  // linearize the input
+  char *buf = memres.allocate_string(size);
+  NgxChainInputStream stream{&chain};
+  auto read = stream.read(reinterpret_cast<std::uint8_t *>(buf), size);
+  if (read < size) {
+    throw std::runtime_error(
+        "mismatch between declared size and read size (read is smaller than "
+        "declared)");
+  }
+
+  if (is_urlencoded(req)) {
+    QueryStringIter it{
+        {buf, size}, memres, '&', QueryStringIter::trim_mode::no_trim};
+
+    // count key occurrences
+    std::unordered_map<std::string_view, std::size_t> bag;
+    for (; !it.ended(); ++it) {
+      std::string_view cur_key = it.cur_key();
+      bag[cur_key]++;
+    }
+
+    // allocate all ddwaf_obj, set keys
+    std::unordered_map<std::string_view, ddwaf_obj *> key_index;
+    ddwaf_map_obj slot_map = slot.make_map(bag.size(), memres);
+    std::size_t i = 0;
+    for (auto &&[key, count] : bag) {
+      ddwaf_obj &cur = slot_map.at_unchecked(i++);
+      key_index.emplace(key, &cur);
+      cur.set_key(key);
+      if (count == 1) {
+        cur.make_string(""sv);  // to be filled later
+      } else {
+        cur.make_array(count, memres);
+        cur.nbEntries = 0;  // fixed later
+      }
+    }
+
+    // set values
+    it.reset();
+    for (it.reset(); !it.ended(); ++it) {
+      auto [cur_key, cur_value] = *it;
+      ddwaf_obj &cur = *key_index.at(cur_key);
+      if (cur.is_string()) {
+        cur.make_string(cur_value);
+      } else {
+        ddwaf_arr_obj &cur_arr = static_cast<ddwaf_arr_obj &>(cur);
+        cur_arr.at_unchecked(cur_arr.nbEntries++).make_string(cur_value);
+      }
+    }
```

Comment:
Could be done in a single pass.

Reply:
I don't see how. On the first pass we're counting total occurrences of a key, which we need to decide the size of the array or if we just need to allocate for a string ddwaf_object. To do it on one single pass, we'd need to do reallocations, which certainly worse in the worst case scenario, though maybe not in the most typical ones.

Reply:
I meant specifically those two loops, the counting loop happens before these two.

Reply:
they're iterating different things. The first loop will do one iteration per each key, whereas the second will do one per key occurrence. I suppose it is possible, to merge the first loop into the second if you had a flag to tell you whether it's the first time the value was seen, the result is more memory and code that's imo less clear.

Reply:
```cpp

    for (it.reset(); !it.ended(); ++it) {
      auto [cur_key, cur_value] = *it;
      auto key_index_it = key_index.find(cur_key);
      if (key_index_it == key_index.end()) {
         ddwaf_obj &cur = slot_map.at_unchecked(i++);
         auto [new_it, res] = key_index.emplace(key, &cur);
         cur.set_key(key);

         auto count = bag[cur_key]; // Assume it's present
         if (count == 1) {
           cur.make_string(""sv);  // to be filled later
         } else {
           cur.make_array(count, memres);
           cur.nbEntries = 0;  // fixed later
         }
         key_index_it = new_it;
      }
      
      ddwaf_obj &cur = key_index_it->second;
      if (cur.is_string()) {
        cur.make_string(cur_value);
      } else {
        ddwaf_arr_obj &cur_arr = static_cast<ddwaf_arr_obj &>(cur);
        cur_arr.at_unchecked(cur_arr.nbEntries++).make_string(cur_value);
      }
    }
```

I haven't checked if it works or not, I just roughly copy-pasted to give you an idea:
- You could even reduce lookups by accessing `key_index` with `operator[]` and checking for `nullptr`,  which would allow you to avoid the extra emplace.
- The initialisation could also set the values instead of creating those temporary empty versions.

 Admittedly it's not much easier to read, but it might be simpler overall.

Reply:
ok but in this case you're assuming I didn't heed your previous recommendation if merging the two hash maps

Reply:
Yes, I didn't account for that.